File: calculator/Display.fs
module Display
open DataTypes


type DataHandler = bool -> bool -> Number list -> Number list option -> unit
type NumberHandler = Number -> unit
type DecimalHandler = unit -> unit
type SignalHandler = bool -> unit
type ErrorHandler = bool -> unit
type ClearHandler = unit -> unit

type Display () as this =
  [<DefaultValue>] val mutable private number: Number list
  [<DefaultValue>] val mutable private decimal: Number list option
  [<DefaultValue>] val mutable private signal: bool
  [<DefaultValue>] val mutable private error: bool

  do
    this.error <- false
    this.signal <- false
    this.number <- List.Empty
    this.decimal <- None

  member public _.Add (value: Number) =
    let updateList (lst: Number list) =
      match lst, value with
      | Zero :: [], Zero -> lst
      | _, _ -> value :: lst

    match this.decimal with
    | Some lst -> this.decimal <- Some (updateList lst)
    | None -> this.number <- updateList this.number

    this.Print ()

  member public _.Set error signal number decimal =
    this.error <- error
    this.signal <- signal
    this.number <- number
    this.decimal <- decimal
    this.Print ()

  member public _.SetDecimal () =
    if this.decimal = None
    then this.decimal <- Some []

  member public _.SetSignal state =
    this.signal <- state

  member public _.SetError state =
    this.error <- state

  member private _.RenderDigit (digit: Number) =
    match digit with
    | One -> "1"
    | Two -> "2"
    | Three -> "3"
    | Four -> "4"
    | Five -> "5"
    | Six -> "6"
    | Seven -> "7"
    | Eigth -> "8"
    | Nine -> "9"
    | Zero -> "0"

  member private _.Print () =
    if this.error
    then printf "\u001B[32m[E] "
    else printf "\u001B[33m"
    if this.signal
    then printf "- "
    let rec ``process`` (list: Number List) =
      match list with
      | current :: remain -> printf "%s" (this.RenderDigit current)
                             ``process`` remain
      | [] -> ()
    this.number
    |> List.rev
    |> ``process``
    match this.decimal with
    | Some lst ->
      printf "."
      lst
      |> List.rev
      |> ``process``
    | None -> ()
    printfn "\u001B[39;49m"

  member public _.Clear () = this.Set false false List.Empty None

File: cwlculator.cs/Display.cs
using Calculator.DataTypes;
using static System.Console;

namespace Calculator;
public class Display(int maxDigitCount) : IDisplay
{
    protected IEnumerable<Number> integral = [ new Number.GhostZero() ];
    protected IEnumerable<Number>? @decimal;
    protected bool isNumberNegative = false;
    protected bool hasError = false;

    public int MaxDigitCount => maxDigitCount;
    public int DigitCount => integral.Count() + @decimal?.Count() ?? 0;

    public virtual void Add(Number n)
    {
        if (DigitCount >= maxDigitCount) return;

        if (@decimal is not null)
        {
            @decimal = @decimal.ElementAt(0) is Number.GhostZero
                     ? ([ n ])
                     : @decimal.Append(n);
        }
        else
        {
            if (integral.ElementAt(0) is not Number.GhostZero || n is not Number.Zero)
                integral = integral.ElementAt(0) is Number.GhostZero
                         ? ([n])
                         : integral.Append(n);
        }
        Print();
    }

    public virtual void AddNumber(IEnumerable<Number> integral, IEnumerable<Number>? @decimal, bool isNegative)
    {
        this.integral = integral.Take(MaxDigitCount);
        this.@decimal = @decimal?.Take(MaxDigitCount - this.integral.Count());
        isNumberNegative = isNegative;
        Print();
    }

    public void SetDecimal()
    {
        if (@decimal is null)
        {
            @decimal = [ new Number.GhostZero() ];
            Print();
        }
    }

    public void SetNegative()
    {
        isNumberNegative = true;
        Print();
    }

    public void SetError()
    {
        hasError = true;
        Print();
    }

    protected virtual string RenderDigit(Number d) => d switch
    {
        Number.GhostZero or Number.Zero => "0",
        Number.One => "1",
        Number.Two => "2",
        Number.Three => "3",
        Number.Four => "4",
        Number.Five => "5",
        Number.Six => "6",
        Number.Seven => "7",
        Number.Eight => "8",
        Number.Nine => "9",
        _ => throw new ArgumentOutOfRangeException($"'{d}' is not a valid digit.", (Exception?)null)
    };

    protected virtual void Print()
    {
        if (hasError)
            Write("\u001B[31m[E] ");
        else
            Write("\u001B[32m");

        if (isNumberNegative)
            Write("-");

        Write(string.Join("", integral.Select(n => RenderDigit(n))));

        if (@decimal is not null)
        {
            Write(".");
            Write(string.Join("", @decimal.Select(n => RenderDigit(n))));
        }

        WriteLine("\u001B[39;49m");
    }

    public virtual void Clear()
    {
        integral = [ new Number.GhostZero() ];
        @decimal = null;
        isNumberNegative = false;
        hasError = false;
    }
}
